File: leetcode2/2medium/노형민/Q2476.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public IList<IList<int>> ClosestNodes(TreeNode root, IList<int> queries)
    {
        List<int> sortedValues = [];
        InOrderTraversal(root, sortedValues);

        List<IList<int>> result = [];

        for (int i = 0; i < queries.Count; i++)
        {
            result.Add(FindClosestNodes(sortedValues, queries[i]));
        }

        return result;
    }

    private IList<int> FindClosestNodes(List<int> sortedValues, int query)
    {
        int left = 0, right = sortedValues.Count - 1;
        int closestLessThan = -1, closestGreaterThan = -1;
        while (left <= right)
        {
            int mid = left + (right - left) / 2;
            if (sortedValues[mid] < query)
            {
                closestLessThan = sortedValues[mid];
                left = mid + 1;
            }
            else if (sortedValues[mid] > query)
            {
                closestGreaterThan = sortedValues[mid];
                right = mid - 1;
            }
            else
            {
                return new List<int> { sortedValues[mid], sortedValues[mid] };
            }
        }
        return [closestLessThan, closestGreaterThan];
    }

    private void InOrderTraversal(TreeNode node, List<int> values)
    {
        if (node == null)
        {
            return;
        }
        InOrderTraversal(node.left, values);
        values.Add(node.val);
        InOrderTraversal(node.right, values);
    }
}

File: leetcode2/2medium/최원준/Q2476.java
package Leetcode.최원준;

/*
1. 아이디어 :


2. 시간복잡도 :
O( )

3. 자료구조/알고리즘 :

 */

public class Q2476 {
/*
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
class Solution {
    TreeSet<Integer> values = new TreeSet<>();

    private void dfs(TreeNode node) {
        if (node==null) return;
        dfs(node.left);
        values.add(node.val);
        dfs(node.right);
    }
    public List<List<Integer>> closestNodes(TreeNode root, List<Integer> queries) {
        dfs(root);
        List<List<Integer>> ans = new ArrayList<>();
        for (int q : queries) {
            List<Integer> qAns = new ArrayList<>();
            Integer left = values.floor(q);
            qAns.add(left == null? -1:left);

            Integer right = values.ceiling(q);
            qAns.add(right == null? -1:right);
            ans.add(qAns);
        }
        return ans;
    }
}
 */
}
